File: The Great Elf Historian Office Mix-Up: Where’s My ID?/puzzle1.py
from typing import List, Tuple

def quick_sort(arr: List[int]) -> List[int]:
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

def parse_input(content: str) -> Tuple[List[str], List[str]]:
    list1 = []
    list2 = []
    for line in content.split("\n"):
        split = line.split("   ") # I don't care if it only works with the input, who needs error handling
        if (len(split) < 2):
            break
        list1.append(split[0])
        list2.append(split[1])
    return list1, list2

def add_distance(list1: List[str], list2: List[str]) -> int:
    total = 0
    for i in range(len(list1)):
        total += abs(int(list1[i]) - int(list2[i]))
    return total

def main() -> None:
    file = open("input.txt", "r")
    content = file.read()
    list1, list2 = parse_input(content)
    list1 = quick_sort(list1)
    list2 = quick_sort(list2)
    print(add_distance(list1, list2))

if __name__ == "__main__":
    main()

File: The Great Elf Historian Office Mix-Up: Where’s My ID?/puzzle1.c
/*
** EPITECH PROJECT, 2024
** AdventOfCode2024 [WSL: Ubuntu]
** File description:
** puzzle1
*/

#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <stdlib.h>

char **str_to_word_array(char *str, char *sep);

int arr_len(char **array)
{
    int i = 0;

    for (; array[i]; i++);
    return i;
}

void free_array(char **array)
{
    for (int i = 0; array[i]; i++)
        free(array[i]);
    free(array);
}

void swap(int *a, int *b)
{
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

int swap_list(int *tab, int start, int end)
{
    int i = start;
    int j = end;
    int pivot = tab[start];

    while (i < j) {
        while (tab[i] <= pivot && i <= end - 1)
            i++;
        while (tab[j] > pivot && j >= start + 1)
            j--;
        if (i < j)
            swap(&tab[i], &tab[j]);
    }
    swap(&tab[start], &tab[j]);
    return j;
}

void quick_sort(int *tab, int start, int end)
{
    int pivot = 0;

    if (start < end) {
        pivot = swap_list(tab, start, end);
        quick_sort(tab, start, pivot - 1);
        quick_sort(tab, pivot + 1, end);
    }
}

int parse_input(char *content)
{
    char **array = str_to_word_array(content, "\n");
    char **line = NULL;
    int *tab1 = malloc(sizeof(int) * (arr_len(array) + 1));
    int *tab2 = malloc(sizeof(int) * (arr_len(array) + 1));
    int i = 0;
    int total = 0;

    for (; array[i]; i++) {
        line = str_to_word_array(array[i], " ");
        tab1[i] = atoi(line[0]);
        tab2[i] = atoi(line[1]);
        free_array(line);
    }
    quick_sort(tab1, 0, arr_len(array) - 1);
    quick_sort(tab2, 0, arr_len(array) - 1);
    for (i = 0; i < arr_len(array); i++) {
        total += abs(tab1[i] - tab2[i]);
    }
    free_array(array);
    free(tab1);
    free(tab2);
    return total;
}

int main(void)
{
    int file = open("input.txt", O_RDONLY);
    struct stat sb;
    char *content;

    if (file == -1)
        return 1;
    if (fstat(file, &sb) == -1)
        return 1;
    content = malloc(sb.st_size);
    if (read(file, content, sb.st_size) != sb.st_size)
        return 1;
    printf("%d\n", parse_input(content));
    free(content);
    close(file);
    return 0;
}